

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>De Genes a Genomas &#8212; Material para el curso de Bioinformatica-EIA</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'genomica';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Hidden Markov Models" href="hmm.html" />
    <link rel="prev" title="Inferencia filogenética y árboles" href="phylo_trees.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="_static/logo.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Curso de Bioinformática
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="intro_linux.html">Introducción a la Terminal</a></li>
<li class="toctree-l1"><a class="reference internal" href="exscripts.html">Scripts Ejecutables</a></li>
<li class="toctree-l1"><a class="reference internal" href="variables.html">Variables, declaraciones y loops</a></li>
<li class="toctree-l1"><a class="reference internal" href="python.html">Introducción a Python y Biopython</a></li>

<li class="toctree-l1"><a class="reference internal" href="Taller1.html">Taller Evaluable</a></li>
<li class="toctree-l1"><a class="reference internal" href="python_seq.html">Manejo de Secuencias con python</a></li>
<li class="toctree-l1"><a class="reference internal" href="alineamientos.html">Alienamientos</a></li>

<li class="toctree-l1"><a class="reference internal" href="ncbi%26blast.html">Herraminetas de NCBI</a></li>


<li class="toctree-l1"><a class="reference internal" href="intro_R.html">Introducción a R</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylo_trees.html">Inferencia filogenética y árboles</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">De Genes a Genomas</a></li>
<li class="toctree-l1"><a class="reference internal" href="hmm.html">Hidden Markov Models</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Fgenomica.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/genomica.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>De Genes a Genomas</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cracking-the-code">“Cracking the code”</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ejemplo-estadisticas-de-dna-completo">Ejemplo: Estadísticas de DNA completo</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="de-genes-a-genomas">
<h1>De Genes a Genomas<a class="headerlink" href="#de-genes-a-genomas" title="Permalink to this heading">#</a></h1>
<p>Haste el momento hemos partido de genes que conocemos y hemos buscado su secuencia y sus homólogos utilizando bases de datos y alineamientos. Sin embargo, cuando secuenciamos un genoma completo no sabemos de antemano cual es su <em>arquitectura</em> (el orden y ubicación de los elementos del genoma).</p>
<section id="cracking-the-code">
<h2>“Cracking the code”<a class="headerlink" href="#cracking-the-code" title="Permalink to this heading">#</a></h2>
<p>Para extraer la información que contiene un genoma desconocido nos basamos en <em>patrones</em> de secuencias conocidas</p>
<section id="ejemplo-estadisticas-de-dna-completo">
<h3>Ejemplo: Estadísticas de DNA completo<a class="headerlink" href="#ejemplo-estadisticas-de-dna-completo" title="Permalink to this heading">#</a></h3>
<p>Algunas características simples de obtener de los genomas son útiles para caracterizarlos:</p>
<ul class="simple">
<li><p>Longitud</p></li>
</ul>
<img src="Genome_Sizes.png"/>
By Abizar at English Wikipedia, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=19537795<ul class="simple">
<li><p>Frecuencia de nucleótidos
La proporción entre A, G, C, and T es una de las propiedades mas fundamentales en la composición de un genoma</p></li>
</ul>
<ul class="simple">
<li><p>Contenido GC</p></li>
</ul>
<img src="cg_genome_cds.webp"/>
https://doi.org/10.1007/s00239-024-10160-5
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Mucha parte de este código está basado en </span>
<span class="c1"># https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter2.html</span>
<span class="c1"># Cargar las librerías</span>
<span class="n">library</span><span class="p">(</span><span class="n">seqinr</span><span class="p">)</span> <span class="c1"># si no lo tiene instalado, instale via install.packages(&quot;Seqinr&quot;)</span>

<span class="c1"># Definir el directorio de trabajo</span>
<span class="n">setwd</span><span class="p">(</span><span class="s2">&quot;/Users/laura/docencia/eafit/comp_bio/2024-2/biocomp_genomica/&quot;</span><span class="p">)</span> 

<span class="c1"># Leer el genoma de covid con función de </span>
<span class="n">covid</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">fasta</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;Sars_cov.dna.fa&quot;</span><span class="p">)</span> <span class="c1">#archivo descargado previamente y copiado en dir de trabajo</span>

<span class="c1"># El primer elemento del objeto es la secuencia</span>
<span class="n">covidseq</span><span class="o">&lt;-</span><span class="n">covid</span><span class="p">[[</span><span class="mi">1</span><span class="p">]]</span>

<span class="c1"># Longitud</span>
<span class="n">length</span><span class="p">(</span><span class="n">covidseq</span><span class="p">)</span>

<span class="c1"># Frecuencia de nucleótidos</span>
<span class="n">table</span><span class="p">(</span><span class="n">covidseq</span><span class="p">)</span>

<span class="c1">#GC content = (number of Gs + number of Cs)*100/(genome length)</span>
<span class="mi">100</span><span class="o">*</span><span class="p">(</span><span class="mi">5492</span><span class="o">+</span><span class="mi">5863</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="mi">8954</span><span class="o">+</span><span class="mi">5492</span><span class="o">+</span><span class="mi">5863</span><span class="o">+</span><span class="mi">9594</span><span class="p">)</span> 
<span class="n">GC</span><span class="p">(</span><span class="n">covidseq</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>El contenido de GC no solo varía entre especies, sino a lo largo del genoma. Lugares que codifican están caracterizados por tener un mayor contenido de GC (respecto al GC de su genoma) que partes no codificadoras. Para observar la variación local en contenido GC podemos fraccionar el genoma en fragmentos de 5000 nucleótidos, lo cual se denomina “sliding window”</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">GC</span><span class="p">(</span><span class="n">covidseq</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">5000</span><span class="p">])</span>
<span class="n">GC</span><span class="p">(</span><span class="n">covidseq</span><span class="p">[</span><span class="mi">5001</span><span class="p">:</span><span class="mi">10000</span><span class="p">])</span>
<span class="n">GC</span><span class="p">(</span><span class="n">covidseq</span><span class="p">[</span><span class="mi">10001</span><span class="p">:</span><span class="mi">15000</span><span class="p">])</span>
<span class="n">GC</span><span class="p">(</span><span class="n">covidseq</span><span class="p">[</span><span class="mi">15001</span><span class="p">:</span><span class="mi">20000</span><span class="p">])</span>
<span class="n">GC</span><span class="p">(</span><span class="n">covidseq</span><span class="p">[</span><span class="mi">20001</span><span class="p">:</span><span class="mi">29903</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<p>En vez de escribir cada función <strong>iterativamente</strong> podemos incluirla en un <em>for-loop</em></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">starts</span> <span class="o">&lt;-</span> <span class="n">seq</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">covidseq</span><span class="p">)</span><span class="o">-</span><span class="mi">5000</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">)</span> <span class="c1"># define una secuencia cada 5000 (ej seq(2,10,by=2))</span>
<span class="n">starts</span>
<span class="n">n</span> <span class="o">&lt;-</span> <span class="n">length</span><span class="p">(</span><span class="n">starts</span><span class="p">)</span>    <span class="c1"># La longitud de este vector define el número de iteraciones</span>
<span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">n</span><span class="p">)</span> <span class="p">{</span>        
  <span class="n">chunk</span> <span class="o">&lt;-</span> <span class="n">covidseq</span><span class="p">[</span><span class="n">starts</span><span class="p">[</span><span class="n">i</span><span class="p">]:(</span><span class="n">starts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="mi">4999</span><span class="p">)]</span> <span class="c1"># fragmento de secuencia</span>
  <span class="n">chunkGC</span> <span class="o">&lt;-</span> <span class="n">GC</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span>                          <span class="c1"># contenido GC en fragmento</span>
  <span class="nb">print</span> <span class="p">(</span><span class="n">chunkGC</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p>Ahora, en vez de imprimir (<code class="docutils literal notranslate"><span class="pre">print</span></code>) el contenido GC en cada iteración, lo vamos a almacenar en un vector para graficarlo.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">chunkGCs</span> <span class="o">&lt;-</span> <span class="n">numeric</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="c1"># Definimos un vector inicial de 0 con la misma longitud del vector &quot;starts&quot;</span>
<span class="c1"># volvemos a iterar en el foor-loop</span>
<span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">n</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">chunk</span> <span class="o">&lt;-</span> <span class="n">dengueseq</span><span class="p">[</span><span class="n">starts</span><span class="p">[</span><span class="n">i</span><span class="p">]:(</span><span class="n">starts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="mi">4999</span><span class="p">)]</span>  <span class="c1"># para un fragmento de nucleótiods de 5000bp &quot;chunk&quot;</span>
        <span class="n">chunkGC</span> <span class="o">&lt;-</span> <span class="n">GC</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span>                            <span class="c1"># hallamos su contenido GC(chunk)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">chunkGC</span><span class="p">)</span>
        <span class="n">chunkGCs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">chunkGC</span>                          <span class="c1"># y lo almacenamos en el vector chunkGCs reemplazando el 0 por el valor corresponidente a i</span>
     <span class="p">}</span>

<span class="c1"># vector de starts</span>
<span class="n">starts</span>
<span class="c1"># vector de contenido GC </span>
<span class="n">chunkGCs</span>
</pre></div>
</div>
</div>
</div>
<p>Ahora vamos a realizar una gráfica que contiene en el eje x a la posición de inicio de la ventana (<code class="docutils literal notranslate"><span class="pre">chunk</span></code>) de DNA y en el eje j el contenido de GC correspondiente <code class="docutils literal notranslate"><span class="pre">chunkGC</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot</span><span class="p">(</span><span class="n">starts</span><span class="p">,</span><span class="n">chunkGCs</span><span class="p">,</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;b&quot;</span><span class="p">,</span><span class="n">xlab</span><span class="o">=</span><span class="s2">&quot;Nucleotide start position&quot;</span><span class="p">,</span><span class="n">ylab</span><span class="o">=</span><span class="s2">&quot;GC content&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Antes de continuar haremos un Intermezzo sobre cómo escribir una función en R personalizada, para luego aplicarla  de nuevo en una gráfica “sliding window” más flexible</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Cómo escribir una función personalizada</span>
<span class="c1"># La función se nombra: &quot;mifuncion&quot;)</span>
<span class="c1"># Se define en terminos de algun(os) parámetro(s) de entrada: x</span>
<span class="c1"># Se escribe la salida &quot;return&quot;: 20 + x*x</span>
<span class="n">mifuncion</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span> <span class="k">return</span><span class="p">(</span><span class="mi">20</span> <span class="o">+</span> <span class="p">(</span><span class="n">x</span><span class="o">*</span><span class="n">x</span><span class="p">))</span> <span class="p">}</span> <span class="c1">#la definición de la función está entre los corchetes</span>
<span class="n">mifuncion</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="n">mifuncion</span><span class="p">(</span><span class="mi">25</span><span class="p">)</span>
<span class="n">mifuncion</span> <span class="c1">#al escribir una función se muestra su definición</span>
</pre></div>
</div>
</div>
</div>
<p>Asimismo podemos definir una gráfica con dos variables (para los ejes x y y) para cualquier tamaño de ventana y cualquier secuencia</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">slidingwindowplot</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">windowsize</span><span class="p">,</span> <span class="n">inputseq</span><span class="p">)</span> <span class="c1"># nombra la función y se definen las variables de entrada </span>
<span class="p">{</span>
   <span class="n">starts</span> <span class="o">&lt;-</span> <span class="n">seq</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">inputseq</span><span class="p">)</span><span class="o">-</span><span class="n">windowsize</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="n">windowsize</span><span class="p">)</span>
   <span class="n">n</span> <span class="o">&lt;-</span> <span class="n">length</span><span class="p">(</span><span class="n">starts</span><span class="p">)</span>    
   <span class="n">chunkGCs</span> <span class="o">&lt;-</span> <span class="n">numeric</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> 
   <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">n</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">chunk</span> <span class="o">&lt;-</span> <span class="n">inputseq</span><span class="p">[</span><span class="n">starts</span><span class="p">[</span><span class="n">i</span><span class="p">]:(</span><span class="n">starts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="n">windowsize</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
        <span class="n">chunkGC</span> <span class="o">&lt;-</span> <span class="n">GC</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">chunkGC</span><span class="p">)</span>
        <span class="n">chunkGCs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">chunkGC</span>
   <span class="p">}</span>
   <span class="n">plot</span><span class="p">(</span><span class="n">starts</span><span class="p">,</span><span class="n">chunkGCs</span><span class="p">,</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;b&quot;</span><span class="p">,</span><span class="n">xlab</span><span class="o">=</span><span class="s2">&quot;Nucleotide start position&quot;</span><span class="p">,</span><span class="n">ylab</span><span class="o">=</span><span class="s2">&quot;GC content&quot;</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># Llamamos la función</span>
<span class="n">slidingwindowplot</span><span class="p">(</span><span class="mi">3000</span><span class="p">,</span> <span class="n">covidseq</span><span class="p">)</span>

<span class="c1"># Podemos guardar esta gráfica en pdf</span>
<span class="n">pdf</span><span class="p">(</span><span class="s2">&quot;covid_sw3000.pdf&quot;</span><span class="p">)</span>
<span class="n">slidingwindowplot</span><span class="p">(</span><span class="mi">3000</span><span class="p">,</span> <span class="n">covidseq</span><span class="p">)</span>
<span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="alert alert-block alert-info">
<b>Ejercicio</b> 
 Despliegue las gráficas de sliding window para una ventana de 500
</div></section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="phylo_trees.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Inferencia filogenética y árboles</p>
      </div>
    </a>
    <a class="right-next"
       href="hmm.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Hidden Markov Models</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cracking-the-code">“Cracking the code”</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ejemplo-estadisticas-de-dna-completo">Ejemplo: Estadísticas de DNA completo</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Laura Salazar Jaramillo
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>