

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Ejercicio de manejo de Secuencias con python &#8212; Material para el curso de Bioinformatica-EIA</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'python_motifs';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="_static/logo.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Curso de Bioinformática
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="intro_linux.html">Introducción a la Terminal</a></li>
<li class="toctree-l1"><a class="reference internal" href="exscripts.html">Scripts Ejecutables</a></li>
<li class="toctree-l1"><a class="reference internal" href="variables.html">Variables, declaraciones y loops</a></li>
<li class="toctree-l1"><a class="reference internal" href="python.html">Introducción a Python y Biopython</a></li>

<li class="toctree-l1"><a class="reference internal" href="Taller1.html">Taller Evaluable</a></li>
<li class="toctree-l1"><a class="reference internal" href="python_seq.html">Manejo de Secuencias con python</a></li>
<li class="toctree-l1"><a class="reference internal" href="alineamientos.html">Alienamientos</a></li>

<li class="toctree-l1"><a class="reference internal" href="ncbi%26blast.html">Herraminetas de NCBI</a></li>


<li class="toctree-l1"><a class="reference internal" href="intro_R.html">Introducción a R</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylo_trees.html">Inferencia filogenética y árboles</a></li>
<li class="toctree-l1"><a class="reference internal" href="genomica.html">De Genes a Genomas</a></li>
<li class="toctree-l1"><a class="reference internal" href="hmm.html">Hidden Markov Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="highthroughput.html">Secuenciación Masiva</a></li>
<li class="toctree-l1"><a class="reference internal" href="AlineamientosCuentas.html">Alineamientos y cuentas</a></li>

</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Fpython_motifs.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/python_motifs.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Ejercicio de manejo de Secuencias con python</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#escribir-secuencias-en-un-archivo">Escribir secuencias en un archivo</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="ejercicio-de-manejo-de-secuencias-con-python">
<h1>Ejercicio de manejo de Secuencias con python<a class="headerlink" href="#ejercicio-de-manejo-de-secuencias-con-python" title="Permalink to this heading">#</a></h1>
<p>Una vez instalados python y biopython, vamos a utilizarlos para manejar las secuencias utilizando el el archivo de covid19 descargado previamente. Para comenzar, vamos s utilizar algunos de los ejemplos de la introducción a linux para entender cómo hacerlos con python. Cada vez que entramos a la terminal debemos activa el entorno de conda que creamos</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">source</span> <span class="n">activate</span> <span class="n">python3</span><span class="mf">.9</span>
<span class="n">python</span>
</pre></div>
</div>
</div>
</div>
<p>En python se pueden buscar patrones de texto utilizando <code class="docutils literal notranslate"><span class="pre">re.search()</span></code>, que es una función (note el parentésis) que busca la primera coincidencia con un patrón especificado. Haremos uso de esta función para busucar patrones de secuencias llamados <em>motifs</em>. Estos son…</p>
<p>En la sesión anterior definimos el codón de inicio y un loop para buscarlo (ver <em>Manejo de secuencias con python</em>). Podemos simplificar esta búsqueda utilizando la función <code class="docutils literal notranslate"><span class="pre">re.search()</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># llamamos las librerías</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">Bio</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="c1"># define una variable para abrir el archivo fasta como input.</span>
<span class="c1"># escribo la ruta completa para no tener problemas localizando el input </span>
<span class="n">dna_file</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;/Users/laura/docencia/eia/bioinfo/clase_python/Sars_cov.dna.fa&#39;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

<span class="c1"># define una variable para la secuencia completa del covid</span>
<span class="n">covid19_seq</span> <span class="o">=</span> <span class="n">dna_file</span><span class="o">.</span><span class="n">seq</span>

<span class="c1"># define la secuencia a buscar</span>
<span class="n">start_codon</span> <span class="o">=</span> <span class="s1">&#39;AUG&#39;</span> 

<span class="c1"># Búsqueda en DNA</span>
<span class="n">covid19_seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">start_codon</span><span class="p">)</span>

<span class="c1"># Búsqueda en el transcrito</span>
<span class="n">covid19_mRNA</span><span class="o">=</span><span class="n">covid19_seq</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="n">start_pos</span><span class="o">=</span><span class="n">covid19_mRNA</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">start_codon</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start_pos</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">covid19_mRNA</span><span class="p">),</span> <span class="mi">3</span><span class="p">):</span>
<span class="c1">#    print(i)</span>
    <span class="n">codon</span> <span class="o">=</span> <span class="n">covid19_mRNA</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="mi">3</span><span class="p">]</span>    
    <span class="k">if</span> <span class="n">codon</span> <span class="o">==</span> <span class="n">start_codon</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">codon</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Ya que un genoma es una secuencia de texto, se puede manejar de forma similar. Para facilitar el código utilizamos la librería de Biopython que ya tiene implementadas funciones específicas para secuencias.</p>
<p>Una vez dentro de python (y cada vez que ingresamos) debemos volver a llamar las librería y definir las variables</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">Bio</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="c1"># define una variable para abrir el archivo fasta como input</span>
<span class="n">dna_file</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;Sars_cov.dna.fa&#39;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

<span class="c1"># define una variable para la secuencia completa del covid</span>
<span class="n">covid19_seq</span> <span class="o">=</span> <span class="n">dna_file</span><span class="o">.</span><span class="n">seq</span>

<span class="c1"># cuenta la longitud del genoma de covid</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;El genoma del virus que causa Covid-19 (SARS-CoV-2) consiste de </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">covid19_seq</span><span class="p">)</span><span class="si">}</span><span class="s1"> bases.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>El comando <code class="docutils literal notranslate"><span class="pre">print</span></code> imprime en la pantalla el texto literal (entre comillas ‘’) y la función <code class="docutils literal notranslate"><span class="pre">len()</span></code> con el parámetro <code class="docutils literal notranslate"><span class="pre">covid19_seq</span></code> que definimos anteriormente.
La <code class="docutils literal notranslate"><span class="pre">f</span></code> dentro de print es un prefijo que formatea la cadena literal (para más <a class="reference external" href="https://docs.python.org/es/3/tutorial/inputoutput.html%5D">información</a>).</p>
<p>En este caso nos da la longitud del genoma de covid que es de 29903 (el genoma humano está en la escala de miles de millones ó GB).</p>
<p>La variable <code class="docutils literal notranslate"><span class="pre">covid19_seq</span></code> es una lista ordenada de bases (string), es decir que la letra de cada posición se puede saber por su <strong>índice</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># La primera posición empieza con el índice 0</span>
<span class="n">covid19_seq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="c1"># Segunda posición</span>
<span class="n">covid19_seq</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="c1"># Las primeras 50 posiciones</span>
<span class="n">covid19_seq</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">50</span><span class="p">]</span>
<span class="c1"># ó</span>
<span class="n">covid19_seq</span><span class="p">[:</span><span class="mi">50</span><span class="p">]</span>
<span class="c1"># Y de atrás para delante</span>
<span class="n">covid19_seq</span><span class="p">[</span><span class="o">-</span><span class="mi">50</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>Asímismo, los condicionales y bucles funcionan similar que en la terminal. Por ejemplo</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
<span class="c1">## Condicionales lógicos</span>
<span class="mi">5</span> <span class="o">==</span> <span class="mi">5</span>
<span class="mi">5</span> <span class="o">==</span> <span class="mi">4</span>
<span class="mi">4</span> <span class="o">&lt;</span> <span class="mi">5</span>

<span class="c1"># El mismo ejemplo de la terminal de imprimir los número de 1-10 con for loop</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

<span class="c1"># El mismo ejemplo de la terminal de imprimir los número de 1-10 con while loop</span>
<span class="n">i</span> <span class="o">=</span> <span class="mi">1</span> 
<span class="k">while</span> <span class="n">i</span><span class="o">&lt;</span><span class="mi">10</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
<p>Ahora vamos a utilizar los bucles y condicionales para buscar el codón de inicio de transcripción para el genoma de covid19</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define el codón de inicio</span>
<span class="n">start_codon</span> <span class="o">=</span> <span class="s1">&#39;AUG&#39;</span>

<span class="c1"># Escanea la secuencia hasta encontrar el patrón</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">covid19_seq</span><span class="p">)):</span>
    <span class="k">if</span> <span class="n">covid19_seq</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="n">start_codon</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;The start codon starts at index&#39;</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
        <span class="k">break</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Codon not found in sequence.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Por qué este script no encontró un codón de inicio? Porque el codón está en términos del RNA mensajero y el input que le dimos fue el del genoma. Para transcribirlo, vamos a definir otra variable</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">covid_mRNA</span> <span class="o">=</span> <span class="n">covid19_seq</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="alert alert-block alert-info">
<b>Ejercicio</b> Vuelva a correr el script anterior para el RNA mensajero y verifique si la posición de indicada si contiene el códon de inicio </div><section id="escribir-secuencias-en-un-archivo">
<h2>Escribir secuencias en un archivo<a class="headerlink" href="#escribir-secuencias-en-un-archivo" title="Permalink to this heading">#</a></h2>
<p>Con frecuencia estamos interesados en un subconjunto de los genes o proteínas para realizar un análisis a profundidad de ellos. En este caso vamos a seleccionar las glicoproteínas que son aquellas que codifican la espícula. Esta espícula (“spike”) es la encargada de mediar la entrada del virus hacia el hospedero y tiene una tasa de mutación mas alta que los otros genes, lo que presumiblemente dió origen a cepas mas virulentas (<a class="reference external" href="https://en.wikipedia.org/wiki/Coronavirus_spike_protein">https://en.wikipedia.org/wiki/Coronavirus_spike_protein</a>). Para esto vamos a llamar al archivo de proteínas dentro de una variable. Recordemos que el archivo se puede llamar sin añadir la ruta completa si abrimos python dentro de la carpeta donde se encuentra el archivo. De lo contrario debemos darle la ruta</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>

<span class="n">prot_file</span> <span class="o">=</span> <span class="s2">&quot;/Users/laura/docencia/eia/bioinfo/clase_python/Sars_cov.prot.fa&quot;</span>

<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">prot_file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">9</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;glycoprotein&quot;</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Cuantas secuencias contiene el archivo? Vamos a imprimir su id, su tamaño y su descripción:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">prot_file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>La descripción tiene mucha información importante pero esta no es muy clara, entonces vamos a extraer el fragmento que contiene la función. Para esto dividimos la descripción en items, utilizando la función <code class="docutils literal notranslate"><span class="pre">split.()</span></code>. Esta función separa una lista en items. Por ejemplo, yo puedo descomponer mi nombre en las letras que lo componen:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="s2">&quot;laura&quot;</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">l</span><span class="o">.</span><span class="n">split</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<p>Asímismo, se puede separar la descripción en las palabras que la componen.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">prot_file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<p>y luego le damos el número  del item de interés <code class="docutils literal notranslate"><span class="pre">[]</span></code> (es el item 9 pero python comienza a contar desde 0, entonces es el 8)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>

<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">prot_file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">9</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;glycoprotein&quot;</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Hasta aquí hemos seleccionado las secuencias que contienen el término “glycoprotein”. Para escribirlas en un archivo en un formato adecuado de secuencias vamos a utilizar la función de biopython  <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.write()</span></code> y a cambiar un poco el estilo.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">glycoproteins</span> <span class="o">=</span> <span class="p">(</span><span class="n">record</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">prot_file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">9</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;glycoprotein&quot;</span><span class="p">))</span>
<span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">glycoproteins</span><span class="p">,</span> <span class="s2">&quot;sars_covid19_glycoproteins.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Donde eperas que esté el archivo? Chequéalo mirando el directorio actual (“get current working directory”). Verifica que si se haya generado con el nombre y formatos correctos, y la información que deseamos que contenga.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>

<span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>
<span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Por último, vamos a escribir los comando que utilizamos en un script de python (con extensión “.py”) para dejar registro de lo que hicimos y pueda ser repetible</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/usr/local/bin/</span>

<span class="c1"># Script por Laura Salazar Jaramillo</span>
<span class="c1"># Genera un subconjunto de glicoproteinas de SARS-covid 19</span>

<span class="c1"># Importamos las librerías necesarias</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="kn">import</span> <span class="nn">Bio</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>

<span class="c1"># Archivos de entrada (inputs)</span>
<span class="n">dna_file</span> <span class="o">=</span> <span class="s2">&quot;/home/lsalazarj/bicomp2024/Sars_cov.dna.fa&quot;</span>
<span class="n">prot_file</span> <span class="o">=</span> <span class="s2">&quot;Sars_cov.prot.fa&quot;</span>


<span class="c1"># Listamos el tamano del genoma</span>
<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">dna_file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;La longitud en nc del genoma es:&quot;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>

<span class="c1"># Listamos los id y longitudes de las proteínas</span>
<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">prot_file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">9</span><span class="p">])</span>

<span class="c1"># Seleccionamos las glicoproteínas y las guardamos en un archivo aparte</span>
<span class="n">glycoproteins</span> <span class="o">=</span> <span class="p">(</span><span class="n">record</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">prot_file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">9</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;glycoprotein&quot;</span><span class="p">))</span>
<span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">glycoproteins</span><span class="p">,</span> <span class="s2">&quot;sars_covid19_glycoproteins.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Este script se puede ejecutar en la terminal escribiendo los comandos</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">glycoproteins</span><span class="o">.</span><span class="n">py</span> <span class="o">&gt;</span> <span class="n">glycoproteins</span><span class="o">.</span><span class="n">out</span>
</pre></div>
</div>
</div>
</div>
<p>Alternativamente se pueden cambiar los permisos del archivo (chmod +x <a class="reference external" href="http://glycoproteins.py">glycoproteins.py</a>) y ejecutarlo</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">./</span><span class="n">glycoproteins</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#escribir-secuencias-en-un-archivo">Escribir secuencias en un archivo</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Laura Salazar Jaramillo
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>